In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [2]:
# View the column names in the data
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price   54639 non-null  float64
 14  unit_cost    54639 non-null  float64
 15  unit_weight  54639 non-null  float64
 16  qty          54639 non-null  int64  
 17  line_number  54639 non-null  int64  
dtypes: float64(3), int64(6), object(9)
memory usag

In [3]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [4]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.head()
# df.sort_values(by='order_week', ascending=False).head()
df.sort_values(by=['category']).head()



first      last                       job                 phone  \
47373     Mark  Thompson  Journalist, broadcasting     529-839-1559x9494   
34204    Angel    Dillon         Market researcher    868.290.8408x61810   
34205  Lindsey  Robinson             Media planner  +1-257-704-0889x8140   
17384   Kristi      West             Herpetologist  +1-790-995-0679x0696   
17383  Rebecca     Brown               Chiropodist      665-955-0156x926   

                           email  client_id  order_id  order_date  order_week  \
47373  markthompson9289@long.com      21314   2266752  2023-01-23           4   
34204   decision.angel@wolf.info      84429   7916403  2023-02-26           8   
34205        robinsonl@allen.com      41561   3527738  2023-03-18          11   
17384         west.bank@beck.org      30827   1001431  2023-01-24           4   
17383       rbrown850@lewis.info      73362   2301885  2023-02-05           5   

       order_year         item_id     category        subcategory  unit_price  \
47373        2023  UXZ38972-80-3D  consumables     paper products       43.32   
34204        2023  MZE18693-10-4D  consumables   kitchen supplies       17.52   
34205        2023  MDA45663-58-3B  consumables   kitchen supplies      333.49   
17384        2023  UCZ01380-66-1U  consumables     paper products       34.82   
17383        2023  ZDA88304-28-6U  consumables  bathroom supplies      119.17   

       unit_cost  unit_weight  qty  line_number  
47373      22.76         3.77   63            9  
34204      13.10         1.90  151            2  
34205     225.79         2.52  437            1  
17384      25.80         2.26   79            7  
17383     100.33         4.15   50            0

In [5]:
# What three item categories had the most entries?
categories = df[['category']].value_counts()
categories.sort_index(ascending=False).head(3)

category   
software        8400
furniture      11915
electronics     2832
Name: count, dtype: int64

In [6]:
# For the category with the most entries, which subcategory had the most entries?
sub_cats = df[['category','subcategory']].value_counts()
sub_cats.sort_index(ascending=False).head(11)

category     subcategory       
software     project management    3522
             logistics             1122
             accounting            3756
furniture    tables                4540
             stands                4201
             seating               2664
             desks                  510
electronics  telecom                978
             computers              467
             cabling                655
             audio visual           732
Name: count, dtype: int64

In [7]:
# Which five clients had the most entries in the data?
best_clients = df[['first','last','client_id']].value_counts(ascending=False)
best_clients.head()

first      last     client_id
Jessica    Reyes    33615        220
Angela     Everett  66037        211
Bryan      Myers    46820        209
Kendra     Garrett  24741        207
Alexandra  Young    38378        207
Name: count, dtype: int64

In [8]:
# Store the client ids of those top 5 clients in a list.
top_five_clients = df[['client_id']].value_counts(ascending=False).head()
top_five_clients

client_id
33615        220
66037        211
46820        209
38378        207
24741        207
Name: count, dtype: int64

In [9]:
# How many total units (the qty column) did the client with the most entries order order?
top_client = df['client_id'].value_counts().idxmax()
total_units = df.loc[df['client_id'] == top_client, 'qty'].sum()
total_units

64313

## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [10]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty


In [11]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.



In [12]:
# Create a column for the total price using the subtotal and the shipping price along with a sales tax of 9.25%


In [13]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).



In [14]:
# Create a column for the profit of each line using line cost and line price



## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [15]:
# Check your work using the totals above


## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [16]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.



In [17]:
# Create a summary DataFrame showing the totals for the for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit. 


In [18]:
# Format the data and rename the columns to names suitable for presentation.

# Define the money columns. 


# Define a function that converts a dollar amount to millions.


# Apply the currency_format_millions function to only the money columns. 


# Rename the columns to reflect the change in the money format. 




In [19]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
